server/src/main/java/org/cloudfoundry/identity/uaa/client/VmidentityClientMetadataProvisioning.java

```diff
@@ -0,0 +1,144 @@
+ *     subcomponents is subject to the terms and conditions of the
+ *     subcomponent's license, as noted in the LICENSE file.
+ *******************************************************************************/
+package org.cloudfoundry.identity.uaa.client;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.cloudfoundry.identity.uaa.oauth.client.ClientConstants;
+import org.cloudfoundry.identity.uaa.util.VmidentityUtils;
+import org.springframework.util.Assert;
+
+import com.vmware.identity.idm.OIDCClient;
+import com.vmware.identity.idm.client.CasIdmClient;
+
+public class VmidentityClientMetadataProvisioning implements ClientMetadataProvisioning {
+
+    private static final Log logger = LogFactory.getLog(VmidentityClientMetadataProvisioning.class);
+
+    private static final String CLIENT_METADATA_SHOW_ON_HOME_PAGE = "show_on_home_page";
+    private static final String CLIENT_METADATA_APP_LAUNCH_URL = "app_launch_url";
+    private static final String CLIENT_METADATA_APP_ICON = "app_icon";
+
+    private CasIdmClient idmClient;
+
+    VmidentityClientMetadataProvisioning(CasIdmClient idmClient) {
+        Assert.notNull(idmClient);
+        this.idmClient = idmClient;
+    }
+
+    @Override
+    public List<ClientMetadata> retrieveAll() {
+        logger.debug("Retrieving UI details for all clients");
+
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            Collection<OIDCClient> clients = idmClient.getOIDCClients(tenant);
+            List<ClientMetadata> clientMetadata = new ArrayList<ClientMetadata>(clients.size());
+            for (OIDCClient client : clients) {
+                clientMetadata.add(mapOIDCClient(client, tenant));
+            }
+            return clientMetadata;
+        } catch (Exception e) {
+            logger.error("Unable to retrieve UI details for all clients", e);
+            throw new VmidentityDataAccessException("Unable to retrieve UI details for all clients");
+        }
+    }
+
+    @Override
+    public ClientMetadata retrieve(String clientId) {
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            OIDCClient client = idmClient.getOIDCClient(tenant, clientId);
+            return mapOIDCClient(client, tenant);
+        } catch (Exception e) {
+            logger.error("Unable to retrieve UI details for client: " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to retrieve UI details for client: " + clientId);
+        }
+    }
+
+    @Override
+    public ClientMetadata update(ClientMetadata resource) {
+        logger.debug("Updating metadata for client: " + resource.getClientId());
```

Comment:
So pretty much everything in these classes, I've tried to copy semantically from their JDBC counterparts. The JDBC variant doesn't require a check here, so my hunch is that the framework will always ensure that this instance at least exists (though it may have empty data).